src/fit.jl

```diff
@@ -433,28 +416,241 @@ function ews_to_abundances(atm, linelist, A_X, measured_EWs, ew_window_size::Rea
+
+    params = [Teff0, logg0, vmic0, metallicity0]
+
+    J_result = DiffResults.JacobianResult(params)
+
+    while true
+        J_result = ForwardDiff.jacobian!(J_result, get_residuals, params)
+        J = DiffResults.jacobian(J_result)
+        residuals = DiffResults.value(J_result)
+        if all((abs.(residuals) .< tolerances)[.! fix_params]) # stopping condition
+            break
         end
+        step = zeros(length(params))
+        step[.! fix_params] = - J[.!fix_params, .!fix_params] \ residuals[.!fix_params]
+        params .+= clamp.(step, -max_step_sizes, max_step_sizes)
+        params .= clamp.(params, first.(parameter_ranges), last.(parameter_ranges))
+    end
+
+    # compute uncertainties
+    stat_σ_r, sys_σ_r = _stellar_param_residual_uncertainties(params, linelist, measured_EWs, measured_EW_err, passed_kwargs)
+    J = DiffResults.jacobian(J_result)
+    stat_σ = abs.(J \ stat_σ_r)
+    sys_σ = abs.(J \ sys_σ_r)
+
+    params, stat_σ, sys_σ
+end
+
+# called by ews_to_stellar_parameters
+function _stellar_param_equation_residuals(params, linelist, EW, EW_err, 
+                                           fix_params, callback, passed_kwargs)
+    A, A_inv_var, neutrals, REWs, Z = 
+        _stellar_param_equations_precalculation(params, linelist, EW, EW_err, passed_kwargs)
+
+
+    teff_residual = _get_slope([line.E_lower for line in linelist[neutrals]],
+                               A[neutrals], A_inv_var[neutrals])
+    logg_residual = (_weighted_mean(A[neutrals], A_inv_var[neutrals]) -
+                     _weighted_mean(A[.! neutrals], A_inv_var[.! neutrals]))
+    vmic_residual = _get_slope(REWs, A[neutrals], A_inv_var[neutrals])
+    feh_residual = _weighted_mean(A, A_inv_var) - (params[4] + Korg.grevesse_2007_solar_abundances[Z])
+    residuals = [teff_residual, logg_residual, vmic_residual, feh_residual]
+    residuals .*= .! fix_params # zero out residuals for fixed parameters
+
+    callback(ForwardDiff.value.(params), ForwardDiff.value.(residuals), ForwardDiff.value.(A))
+    residuals
+end
+
+# called by _stellar_param_equation_residuals
+# returns (statistical_uncertainty, systematic_uncertainty)
+function _stellar_param_residual_uncertainties(params, linelist, EW, EW_err, passed_kwargs)
+    A, A_inv_var, neutrals, REWs, _ = 
+        _stellar_param_equations_precalculation(params, linelist, EW, EW_err, passed_kwargs)
+
+    # estimated total (including systematic) err in the abundances of each line
+    total_err = std(A)
+    total_ivar = ones(length(A)) * total_err^-2
+
+    stat_sigma, total_sigma = map([A_inv_var, total_ivar]) do ivar
+        sigma_mean = 1 ./ sqrt(sum(ivar))
+        teff_residual_sigma = _get_slope_uncertainty([line.E_lower for line in linelist[neutrals]], ivar[neutrals])
+        vmic_residual_sigma = _get_slope_uncertainty(REWs, ivar[neutrals])
+        [teff_residual_sigma, sigma_mean, vmic_residual_sigma, sigma_mean]
+    end
+
+    if EW_err == ones(length(EW))
```

Comment:
I still haven't fully digested how I feel about the systematic uncertainty estimates as we discussed, but I don't think it's wrong. I think it is worth including it in this merge and then thinking more about it.